[Experimental] Expose roll angle of Apple Pencil for `twist` Pointer Event attribute

lilyspiniolas opened the pull request:
#### ef6f17c72eedaaaab06c75497fa417c6c0b6cc45
<pre>
[Experimental] Expose roll angle of Apple Pencil for `twist` Pointer Event attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=298276">https://bugs.webkit.org/show_bug.cgi?id=298276</a>
<a href="https://rdar.apple.com/159705821">rdar://159705821</a>

Reviewed by Abrar Rahman Protyasha.

Expose `rollAngle` of `UITouch` as the value for pointer event property
`twist`. This feature is disabled by default.

Added layout tests to ensure the default value of `twist` remains 0
for pointer events originating from mouse presses, stylus taps, and
taps.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/pointerevents/ios/pointer-event-twist-on-stylus-down-expected.txt: Added.
* LayoutTests/pointerevents/ios/pointer-event-twist-on-stylus-down.html: Added.
* LayoutTests/pointerevents/pointer-event-constructors-twist-expected.txt: Added.
* LayoutTests/pointerevents/pointer-event-constructors-twist.html: Added.
* LayoutTests/pointerevents/pointer-event-twist-on-tap-or-click-expected.txt: Added.
* LayoutTests/pointerevents/pointer-event-twist-on-tap-or-click.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/Touch.cpp:
(WebCore::Touch::Touch):
(WebCore::Touch::cloneWithNewTarget const):
* Source/WebCore/dom/Touch.h:
(WebCore::Touch::create):
(WebCore::Touch::twist const):
* Source/WebCore/dom/ios/PointerEventIOS.cpp:
(WebCore::m_predictedEvents):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::SyntheticTouchPoint::SyntheticTouchPoint):
* Source/WebCore/platform/PlatformTouchPoint.h:
(WebCore::PlatformTouchPoint::twist const):
* Source/WebKit/Shared/WebEvent.serialization.in:
* Source/WebKit/Shared/WebEventConversion.cpp:
(WebKit::WebKit2PlatformTouchPoint::WebKit2PlatformTouchPoint):
* Source/WebKit/Shared/WebTouchEvent.h:
(WebKit::WebPlatformTouchPoint::WebPlatformTouchPoint):
(WebKit::WebPlatformTouchPoint::setTwist):
(WebKit::WebPlatformTouchPoint::twist const):
* Source/WebKit/Shared/ios/NativeWebTouchEventIOS.mm:
(WebKit::NativeWebTouchEvent::extractWebTouchPoints):
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _shouldExposeRollAngleAsTwist]):
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.h:
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.mm:
(-[WKTouchEventsGestureRecognizer _touchEventForChildTouch:withParent:]):
(-[WKTouchEventsGestureRecognizer _recordTouches:ofType:forEvent:]):
* Tools/TestWebKitAPI/Tests/ios/TouchEventTests.mm:

Canonical link: <a href="https://commits.webkit.org/299701@main">https://commits.webkit.org/299701@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04eb77044231926f91791e4730fe4bd42a1939cd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119885 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39578 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30229 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126201 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71958 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/52728264-2eb2-4646-98a9-c2cd49d8b3ec) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121761 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40273 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48155 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91038 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60345 "Passed tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0e3c436f-753e-4be0-b84c-777359e7ec6d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122837 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32158 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107494 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71594 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6de7bad1-3762-4a51-87a6-9cfbdd848a5b) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31188 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25600 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69846 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/112010 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101630 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25790 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129131 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/118401 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46805 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35477 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99659 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47171 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103685 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99504 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25262 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44947 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22962 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43411 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46667 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52373 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/147099 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46133 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37800 "Found 1 new JSC binary failure: testapi, Found 18685 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/push1.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/ControlFlow/forInPrimitive.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/EH/try2.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default ... (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49482 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47819 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/dom/ios/PointerEventIOS.cpp:
```diff
@@ -84,6 +84,12 @@ PointerEvent::PointerEvent(const AtomString& type, const PlatformTouchEvent& eve
+    // The rotation angle is currently in the range
+    // [-π, π). Convert it to range [0°, 359°]
+    // and flip the direction to clockwise.
+    const auto twistInDegrees = 180 / std::numbers::pi * event.twistAtIndex(index);
```
We have `rad2deg()` in `wtf/MathExtras.h` that you could re-use.

aprotyas commented on Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.mm:
```diff
@@ -239,6 +240,18 @@ static CGPoint mapRootViewToViewport(CGPoint pointInRootView, WKContentView *con
+    if (!contentView || !contentView.get()->_page)
+        return NO;
+
+    return contentView.get()->_page->preferences().exposeRollAngleAsTwistEnabled();
```
Might be neater to expose something like this instead of reaching into content view guts from here:

```objc
// WKContentViewInteraction
- (BOOL)shouldExposeRollAngleAsTwist
{
    return _page->preferences().exposeRollAngleAsTwistEnabled();
}
```

aprotyas commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -2759,6 +2759,20 @@ ExposeDefaultSpeakerAsSpecificDeviceEnabled:
 
+ExposeRollAngleAsTwistEnabled:
+  type: bool
+  status: stable
```
nit: Stable doesn't seem like the most appropriate category, I would've gone with Internal or Testable?

pxlcoder commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -2759,6 +2759,20 @@ ExposeDefaultSpeakerAsSpecificDeviceEnabled:
     WebCore:
       default: true
 
+ExposeRollAngleAsTwistEnabled:
```
This needs to be guarded with a HAVE/ENABLE/PLATFORM so it doesn’t show up on Mac.

aprotyas commented on LayoutTests/pointerevents/ios/pointer-event-twist-on-stylus-down.html:
```diff
@@ -0,0 +1,21 @@
+            testRunner.notifyDone();
+        });
+
+        UIHelper.stylusTapAtPoint(1, 1);
```
I think you need to `await` here, also, probably enable the `ExposeRollAngleAsTwistEnabled` for this test?

```html
 <!-- webkit-test-runner [ ExposeRollAngleAsTwistEnabled=true ] -->
```

lilyspiniolas replied:
The test is just making sure that the default twist value hasn't changed from 0, so the setting doesn't need to be enabled.

lilyspiniolas replied:
Also, `await` shouldn't be needed here since the test will not progress until the event listener fires.

lilyspiniolas committed:
[Experimental] Expose roll angle of Apple Pencil for `twist` Pointer Event attribute
https://bugs.webkit.org/show_bug.cgi?id=298276
rdar://159705821

Reviewed by Abrar Rahman Protyasha.

Expose `rollAngle` of `UITouch` as the value for pointer event property
`twist`. This feature is disabled by default.

Added layout tests to ensure the default value of `twist` remains 0
for pointer events originating from mouse presses, stylus taps, and
taps.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/pointerevents/ios/pointer-event-twist-on-stylus-down-expected.txt: Added.
* LayoutTests/pointerevents/ios/pointer-event-twist-on-stylus-down.html: Added.
* LayoutTests/pointerevents/pointer-event-constructors-twist-expected.txt: Added.
* LayoutTests/pointerevents/pointer-event-constructors-twist.html: Added.
* LayoutTests/pointerevents/pointer-event-twist-on-tap-or-click-expected.txt: Added.
* LayoutTests/pointerevents/pointer-event-twist-on-tap-or-click.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/Touch.cpp:
(WebCore::Touch::Touch):
(WebCore::Touch::cloneWithNewTarget const):
* Source/WebCore/dom/Touch.h:
(WebCore::Touch::create):
(WebCore::Touch::twist const):
* Source/WebCore/dom/ios/PointerEventIOS.cpp:
(WebCore::m_predictedEvents):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::SyntheticTouchPoint::SyntheticTouchPoint):
* Source/WebCore/platform/PlatformTouchPoint.h:
(WebCore::PlatformTouchPoint::twist const):
* Source/WebKit/Shared/WebEvent.serialization.in:
* Source/WebKit/Shared/WebEventConversion.cpp:
(WebKit::WebKit2PlatformTouchPoint::WebKit2PlatformTouchPoint):
* Source/WebKit/Shared/WebTouchEvent.h:
(WebKit::WebPlatformTouchPoint::WebPlatformTouchPoint):
(WebKit::WebPlatformTouchPoint::setTwist):
(WebKit::WebPlatformTouchPoint::twist const):
* Source/WebKit/Shared/ios/NativeWebTouchEventIOS.mm:
(WebKit::NativeWebTouchEvent::extractWebTouchPoints):
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _shouldExposeRollAngleAsTwist]):
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.h:
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.mm:
(-[WKTouchEventsGestureRecognizer _touchEventForChildTouch:withParent:]):
(-[WKTouchEventsGestureRecognizer _recordTouches:ofType:forEvent:]):
* Tools/TestWebKitAPI/Tests/ios/TouchEventTests.mm:

Canonical link: https://commits.webkit.org/299701@main
